Introduce xml support (schema and file validation)

jschaedl opened the pull request:
### What was done?

- introduced xml support
- introduced `Schema\Validator` and `File\Validator` factories to use them in `Validation\Provider` class
- added new gherkin **@Then response should be a valid XML** in `ArachneContext`
- added seromenho/xml-validator library
- seperated examples by json and xml
- updated travis-ci config

jschaedl committed:
Introduce xml support (schema and file validation)

- introduced xml support
- introduced `Schema\Validator` and `File\Validator` factories to use them in `Validation\Provider` class
- added new gherkin **@Then response should be a valid XML** in `ArachneContext`
- added seromenho/xml-validator library
- seperated examples by json and xml
- updated travis-ci config

jschaedl commented on examples/xml/behat.yml:
```diff
@@ -0,0 +1,18 @@
+default:
+  extensions:
+    Arachne\ServiceContainer\ArachneExtension:
+      base_url: http://urlecho.appspot.com
```
Because http://echo.jsontest.com/ only echo json, I devided the example behat configuration into json and xml and used a xml echo service.

jschaedl commented on examples/xml/features/example.feature:
```diff
@@ -0,0 +1,34 @@
+    And I send the request
+    Then the status code should be 200
+    And response should be a valid XML
+    And response should validate against "one_two.xml" schema
```
With entering the filename and extension, it is now possible to have a proper schema validation. See `Schema\ValidatorFactory`.

jschaedl commented on examples/xml/features/example.feature:
```diff
@@ -0,0 +1,34 @@
+    Then the status code should be 200
+    And response should be a valid XML
+    And response should validate against "one_two.xml" schema
+    And response should be identical to "one_two.xml" file
```
With entering the filename and extension, it is now possible to have a proper file validation. See `File\ValidatorFactory`.

jschaedl committed:
Change file extension of xml schema files to xsd

jschaedl commented on examples/xml/features/example.feature:
```diff
@@ -0,0 +1,34 @@
+    Then the status code should be 200
+    And response should be a valid XML
+    And response should validate against "one_two.xsd" schema
+    And response should be identical to "one_two.xml" file
```
With entering the filename and extension, it is now possible to have a proper file validation. See `File\ValidatorFactory`.

jschaedl commented on examples/xml/features/example.feature:
```diff
@@ -0,0 +1,34 @@
+    And I send the request
+    Then the status code should be 200
+    And response should be a valid XML
+    And response should validate against "one_two.xsd" schema
```
With entering the filename and extension, it is now possible to have a proper schema validation. See `Schema\ValidatorFactory`.

theDisco commented on src/Arachne/Context/ArachneContext.php:
```diff
@@ -231,6 +231,27 @@ public function responseShouldBeAValidJson()
+        libxml_use_internal_errors(true);
+
+        $xmlDocument = simplexml_load_string($this->getResponse()->getBody());
+
```
Please initialize the `$errors` variable.

jschaedl replied:
ofc

theDisco commented on src/Arachne/Validation/File/ValidatorFactory.php:
```diff
@@ -0,0 +1,43 @@
+     * @return JsonFile|XmlFile
+     */
+    public function create($fileType)
+    {
```
By forcing the users to use the file extension in the scenario you break the backwards compatibility. My initial idea was to use a configuration variable to specify the data type returned by a specific web service under test. In my opinion it is pretty uncommon that the same web service will return two different data types, however. I like this implementation and I would keep it for now, but I would like you to make a small change to this factory and make it always return a `JsonFile` validator, if the file type was not recognized.

jschaedl replied:
Right now I'm in a situation where I need to test both data types in one application. I think it is quite common if you work on an application which integrates with different services/webhooks which provide and expect data in different formats. So I thought it would be a good idea to add the file type recognition to the `ArachneContext` class so that it can be used by all steps. 

Of course you are right my implementation broke backwards compatibility. Which is now ensured :-)

theDisco commented on examples/xml/behat.yml:
```diff
@@ -0,0 +1,18 @@
+      headers:
+        Authorization: Token token=123456
+  suites:
+    json:
```
This is my OCD kicking in, since it is an xml config file, can you change the suite name :)

jschaedl replied:
Absolutely :-)

theDisco reviewed (changes requested):
Thanks Jan for this solid implementation of the XML support in Arachne. Can you please take a look at the comments I made and consider small changes to your implementation.

The background is that I didn't want to make the validation type explicit based on the file extension but rather implicit based on the type of the web service configured in the configuration file. Nevertheless I want to integrate your changes and allow the user to overwrite the validator for the configured data type by providing an extensions to the file. I can take care of providing the implementation of the configurable data type. 

Please provide a test that will make sure that the code is backwards compatible. Can you also add the documentation for the steps you have added?

jschaedl committed:
Introduce requested changes after code review

- returned a JsonFile validator in ValidatorFactory, if the file type was not recognized
- reverted FileLocator to use filename and extension again
- changed Provider and ArachneContext to ensure BC
- added test to ensure BC

jschaedl commented:
Thanks Wojtek for the review, feedback and considering to merge this PR :-)

jschaedl committed:
Adjust README with xml support explanations

theDisco commented:
Thanks again for considering the changes, everything looks good now. I will merge it into master and create a new tag.
